posts/2018-01-26-announcing-pidgin-v2.0.md

```diff
@@ -0,0 +1,304 @@
+
+```csharp
+interface ITokenStream<TToken>
+{
+    void ReadInto(TToken[] buffer, int start, int count);
+}
+```
+
+Why not design this signature as `void ReadInto(Span<TToken> span)`? I'd love to! Sadly that would preclude some implementations of `ITokenStream`. Methods like `stream.Read(Span<byte> span)` are available only on .NET Core, and as far as I'm aware Microsoft has no plans to backport them, so for compatibility with the desktop framework I'm stuck using arrays. This is irksome, as [I've said before](/posts/2018-12-06-zooming-in-on-field-accessors.html) --- what's the point of designing a feature for library authors if you're not going to support it properly? It's obvious that the desktop framework is reaching the end of its life. Microsoft should just grasp the nettle and admit it.
+
+`ReadInto` allows the `ParseState` to fill its buffer in chunks. But `ParseState`'s interface can also be chunk-ified --- some `Parser`s (like `String`) can predict how many characters they'll pull from the input. I added a `Peek` method to `ParseState` which returns a view into the `ParseState`'s buffer. (This makes `Peek` a little tricky to use correctly --- you have to be careful not to continue using the `Span` after a call to `Advance`, which may mutate the buffer.)
+
+```csharp
+public ReadOnlySpan<TToken> Peek(int count);
+```
+
+Using `Peek`, a parser like `String("foo")` can now look at three characters from the input to see if they match `foo`, rather than one at a time. This runs about 20% faster for long strings.
+
+
+Make Your Own Opportunities
+---------------------------
+
+Indulge me for a moment while I dispense some unsolicited career advice. For a long time, performance engineering was something that [other](https://twitter.com/marcgravell/), [cleverer](https://twitter.com/davidfowl) [people](https://mattwarren.org/) did while I observed from a distance with awe. Why is ASP.NET fast? Because the guys that wrote it are Actual Wizards.
+
+I've been lucky enough to work with some of those clever people for a few years now, and actually they're not Actual Wizards. They just have more experience than me --- they know how to use and interpret profiles and benchmarks, they've seen enough to know what does and doesn't work, and they've had practice coming up with ideas to improve performance.
+
+The good news is that it's never too late or too early to start building your own experience level, no matter where you are in your career. For me that meant setting myself the goal of making my open source libraries as fast as I could. Working on performance-focused libraries has been hugely rewarding for me, and learning how to write low-level C# has paid off at work. For you, it could mean picking a slow part of your codebase at work and trying to make it faster, or pair-programming with a colleague, or thoroughly reading through an existing highly optimised codebase. Making your own opportunities to learn new things is the best way to get better.
```

Comment:
+1 might be worth mentioning that a lot of performance work doesn't even need to be this low-level. Being aware of where you're doing IO (DB calls, caching, etc.) and an understanding of different data structures can make a huge difference!

Reply:
Good idea, thanks!